[YowSiaoKang] iP

YowSiaoKang opened the pull request:
### DukePro

>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

 - [x] Managing tasks
 - [x] Managing deadlines (coming soon)
 - [x] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

YowSiaoKang committed:
updated to Rename, Greet, Exit

YowSiaoKang committed:
implemented Echo

YowSiaoKang committed:
implement Add, List

YowSiaoKang committed:
added task class

YowSiaoKang committed:
added ToDos, Events, Deadlines

YowSiaoKang committed:
Automated Text UI Testing

YowSiaoKang committed:
added some Exceptions

YowSiaoKang committed:
implemented Delete function

YowSiaoKang committed:
added enum Command

YowSiaoKang committed:
add Data functions

chewbum commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,174 @@
+import java.util.ArrayList;
+
 public class Duke {
+    private static final String logo = " ____        _        \n"
```
Constants should be capitalized

chewbum commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,174 @@
+    }
+
+    private void list() {
+        System.out.println("\tHere are the tasks in your list:");
+        for (int i = 0; i < this.list.size(); i++) {
+            System.out.println("\t" + (i + 1) + ". " + this.list.get(i));
+        }
+    }
+
+    private void mark(String[] input) throws IncompleteCommandException, InvalidArgumentException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        try {
+            int index = Integer.parseInt(input[1]) - 1;
+            this.list.get(index).markAsDone();
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidArgumentException();
+        }
+    }
+
+    private void unmark(String[] input) throws IncompleteCommandException, InvalidArgumentException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        try {
+            int index = Integer.parseInt(input[1]) - 1;
+            this.list.get(index).markNotDone();
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidArgumentException();
+        }
+    }
+
+    private void addTodo(String[] input) throws IncompleteCommandException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        String description = input[1];
+        this.add(new Todo(description));
+    }
```
Can be abstracted into new packages and classes

chewbum commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,174 @@
+
+    private void addDeadline(String[] input) throws IncompleteCommandException, NoDeadlineException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        input = input[1].split(" /by ", 2);
+        if (input.length == 1) {
+            throw new NoDeadlineException();
+        }
+        String description = input[0];
+        String by = input[1];
+        this.add(new Deadline(description, by));
+    }
+
+    private void addEvent(String[] input) throws IncompleteCommandException, NoPeriodException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        input = input[1].split(" /from ", 2);
+        if (input.length == 1) {
+            throw new NoPeriodException();
+        }
+        String description = input[0];
+        String[] arr = input[1].split(" /to ", 2);
+        if (arr.length == 1) {
+            throw new NoPeriodException();
+        }
+        this.add(new Event(description, arr[0], arr[1]));
+    }
+
+    private void delete(String[] input) throws IncompleteCommandException, InvalidArgumentException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        try {
+            int index = Integer.parseInt(input[1]) - 1;
+            Task task = this.list.get(index);
+            this.list.remove(index);
+            System.out.println("\tNoted. I've removed this task:");
+            System.out.println("\t  " + task);
```
JavaDoc comments could be added to provide more context

chewbum commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,174 @@
+                String userInput = duke.sc.nextLine();
+                String[] input = userInput.split(" ", 2);
+                String action = input[0].toUpperCase();
+                System.out.println(LINE);
+                switch (Command.valueOf(action)) {
+                    case BYE:
+                        duke.sayBye();
+                        return;
+                    case LIST:
+                        duke.list();
+                        break;
+                    case MARK:
+                        duke.mark(input);
+                        break;
+                    case UNMARK:
+                        duke.unmark(input);
+                        break;
+                    case TODO:
+                        duke.addTodo(input);
+                        break;
+                    case DEADLINE:
+                        duke.addDeadline(input);
+                        break;
+                    case EVENT:
+                        duke.addEvent(input);
+                        break;
+                    case DELETE:
+                        duke.delete(input);
+                        break;
+                    default:
+                        throw new UnknownCommandException();
+                }
+            } catch (IllegalArgumentException e) {
+                System.out.println("\t" + new UnknownCommandException());
+            } catch (Exception e) {
+                System.out.println("\t" + e);
+            }
+            System.out.println(LINE);
+        }
     }
```
Good formatting for your loop!

YowSiaoKang committed:
Update file paths to be OS-independent

YowSiaoKang committed:
Handle corrupted data file

YowSiaoKang committed:
Merge branch 'branch-level-7'

YowSiaoKang committed:
Implement LocalDateTIme

YowSiaoKang committed:
Merge branch 'branch-Level-8'

zhikaiong2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,260 @@
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE,
+    }
+
+    private static final String filePath = Paths.get("data", "duke.txt").toString();
```
`filepath` should be in caps(`FILEPATH`) since it is a constant

zhikaiong2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,260 @@
+                        duke.sayBye();
+                        duke.saveData();
+                        return;
+                    case LIST:
+                        duke.list();
+                        break;
+                    case MARK:
+                        duke.mark(input);
+                        break;
+                    case UNMARK:
+                        duke.unmark(input);
+                        break;
+                    case TODO:
+                        duke.addTodo(input);
+                        break;
+                    case DEADLINE:
+                        duke.addDeadline(input);
+                        break;
+                    case EVENT:
+                        duke.addEvent(input);
+                        break;
+                    case DELETE:
+                        duke.delete(input);
+                        break;
+                    default:
+                        throw new UnknownCommandException();
+                }
+            } catch (IllegalArgumentException e) {
+                if (e.getMessage().contains("No enum constant Duke.Command.")) {
+                    System.out.println("\t " + new UnknownCommandException());
+                } else {
+                    System.out.println("\t" + e.getMessage());
+                }
+            } catch (Exception e) {
+                System.out.println("\t" + e);
+            }
+            System.out.println(LINE);
+        }
     }
 }
```
For the switch statement, `case` clauses should not be indented as specified in the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html#layout)

zhikaiong2001 reviewed (commented):
LGTM. Just some comments regarding code standard

zhikaiong2001 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    private DateTimeFormatter formatter = DateTimeFormatter.ofPattern("[yyyy-MM-dd HH:mm][yyyy/MM/dd HH:mm][yyyy MM dd HH:mm][yyyy.MM.dd HH:mm]");
+    private LocalDateTime by;
```
Good job using java.time to handle event and deadline tasks

zhikaiong2001 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,260 @@
+        System.out.println(input);
+    }
+
+    private void add(Task task) {
+        this.list.add(task);
+        System.out.println("\tGot it. I've added this task:");
+        System.out.println("\t  " + task);
+        System.out.println("\tNow you have " + this.list.size() + " tasks in the list.");
+    }
+
+    private void list() {
+        System.out.println("\tHere are the tasks in your list:");
+        for (int i = 0; i < this.list.size(); i++) {
+            System.out.println("\t" + (i + 1) + ". " + this.list.get(i));
+        }
+    }
+
+    private void mark(String[] input) throws IncompleteCommandException, InvalidArgumentException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        try {
+            int index = Integer.parseInt(input[1]) - 1;
+            this.list.get(index).markAsDone();
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidArgumentException();
+        }
+    }
+
+    private void unmark(String[] input) throws IncompleteCommandException, InvalidArgumentException {
+        if (input.length == 1 || input[1].equals("")) {
+            throw new IncompleteCommandException(input[0]);
+        } 
+        try {
+            int index = Integer.parseInt(input[1]) - 1;
+            this.list.get(index).markNotDone();
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidArgumentException();
+        }
+    }
```
Method names could be more descriptive, for example 
`private void addTask(Task task)`
`private void listAllTasks()`

zhikaiong2001 reviewed (commented):
Add on to previous review. Few things to keep in mind,

1.  Constants should be in caps
2.  Format IDE to not indent for `case` clauses
3.  Try to be more descriptive when it comes to naming methods

Overall, good job in making the program work!

YowSiaoKang committed:
Add class Ui

YowSiaoKang committed:
Add class Storage

YowSiaoKang committed:
Add Parser, Tasklist, Command classes

YowSiaoKang committed:
Organize into Packages

YowSiaoKang committed:
Merge remote-tracking branch 'origin/add-gradle-support'

YowSiaoKang committed:
Use Gradle

YowSiaoKang committed:
Add JUnit Tests

YowSiaoKang committed:
Add JavaDoc comments

YowSiaoKang committed:
Follow the Coding Standard

YowSiaoKang committed:
Add find command

YowSiaoKang committed:
Merge branch 'branch-A-JavaDoc'

YowSiaoKang committed:
Merge branch 'branch-A-CodingStandard'

YowSiaoKang committed:
Merge branch 'branch-Level-9'

YowSiaoKang committed:
Use checkStyle

YowSiaoKang committed:
Merge branch 'branch-A-CheckStyle'

YowSiaoKang committed:
Setup JavaFX

YowSiaoKang committed:
Add GUI to the chatbot

YowSiaoKang committed:
Merge branch 'branch-Level-10'

ashleyclx commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,167 @@
+     * @param task Task to show the user.
+     * @return The unmarked task to be shown to the user.
+     */
+    public String showUnmarkTask(Task task) {
```
Since other instances of the word unmark for e.g. in `TaskList` class, you treated it as 2 separate words i.e.
``` 
    public Task unMark(int index) {
        Task task = this.list.get(index);
        task.markNotDone();
        return task;
    }
```
I think it would be more consistent to name it as `showUnMarkTask`?

ashleyclx commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,97 @@
+            description = input[1];
+            return new FindCommand(description);
+        default:
+            throw new UnknownCommandException();
```
Great job! The code readability is really good!

ashleyclx commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,97 @@
+ */
+public class Parser {
+    /**
+     * Parses the user input into a command.
+     * @param command User input.
```
There should be an empty line between your description and parameters

ashleyclx commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,97 @@
+import command.FindCommand;
+import command.ListCommand;
+import command.MarkCommand;
+import command.UnMarkCommand;
+import exception.IncompleteCommandException;
```
It would look nicer to split the command and exception imports for readability

ashleyclx commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,167 @@
+ */
+public class Ui {
+    private final String LINE = "\t____________________________________________________________";
+    private Scanner sc = new Scanner(System.in);
```
Can ignore if you plan to change but what about making `sc` final as well?

ashleyclx reviewed (commented):
LGTM! Overall very good quality so the nits were very nitty and aren't really that important HAHA

YowSiaoKang committed:
Use Assertions

Add assertions to validate data integrity

The code does not have sufficient error checking mechanisms in place. This lack of assertions could lead to undetected bugs or inconsistencies in our data.

Let's add assertions in critical areas of Duke and MainWindow.

YowSiaoKang committed:
Improve Code Quality

Improve code quality for better maintainability

The codebase has some areas that could be improved in terms of readability and maintainability. This can lead to difficulty in understanding the code and potential bugs in the future.

Let's, refactor the code to adhere to best practices. This includes simplifying complex functions, removing redundant code, and improving variable naming.

YowSiaoKang committed:
Merge pull request #1 from YowSiaoKang/branch-A-Assertions

Use Assertions

YowSiaoKang committed:
Merge branch 'master' into branch-A-CodeQuality

YowSiaoKang committed:
Merge pull request #2 from YowSiaoKang/branch-A-CodeQuality

Improve Code Quality

YowSiaoKang committed:
Set up CI

YowSiaoKang committed:
Add Javadoc comment

YowSiaoKang committed:
Implement archiving tasks

YowSiaoKang committed:
Add Javadocs for ArchiveCommand

YowSiaoKang committed:
Merge branch 'branch-C-Archive'

YowSiaoKang committed:
Remove traces of Duke

YowSiaoKang committed:
Improve code quality

YowSiaoKang committed:
Add GUI screenshot

YowSiaoKang committed:
Fix checkstyle

YowSiaoKang committed:
Update README.md
